storm-core/src/jvm/org/apache/storm/daemon/supervisor/workermanager/DefaultWorkerManager.java

```diff
@@ -0,0 +1,408 @@
+        String resourceRoot = stormRoot + Utils.FILE_PATH_SEPARATOR + ConfigUtils.RESOURCES_SUBDIR;
+        String os = System.getProperty("os.name").replaceAll("\\s+", "_");
+        String arch = System.getProperty("os.arch");
+        String archResourceRoot = resourceRoot + Utils.FILE_PATH_SEPARATOR + os + "-" + arch;
+        String ret = archResourceRoot + Utils.CLASS_PATH_SEPARATOR + resourceRoot + Utils.CLASS_PATH_SEPARATOR + conf.get(Config.JAVA_LIBRARY_PATH);
+        return ret;
+    }
+
+    protected String getWorkerClassPath(String stormJar, Map stormConf) {
+        List<String> topoClasspath = new ArrayList<>();
+        Object object = stormConf.get(Config.TOPOLOGY_CLASSPATH);
+
+        if (object instanceof List) {
+            topoClasspath.addAll((List<String>) object);
+        } else if (object instanceof String) {
+            topoClasspath.add((String) object);
+        } else {
+            LOG.error("topology specific classpath is invaild");
+        }
+        String classPath = Utils.workerClasspath();
+        String classAddPath = Utils.addToClasspath(classPath, Arrays.asList(stormJar));
+        return Utils.addToClasspath(classAddPath, topoClasspath);
+    }
+
+    /**
+     * "Generates runtime childopts by replacing keys with topology-id, worker-id, port, mem-onheap"
+     *
+     * @param value
+     * @param workerId
+     * @param stormId
+     * @param port
+     * @param memOnheap
+     */
+    public List<String> substituteChildopts(Object value, String workerId, String stormId, Long port, int memOnheap) {
+        List<String> rets = new ArrayList<>();
+        if (value instanceof String) {
+            String string = (String) value;
+            if (StringUtils.isNotBlank(string)){
+                string = string.replace("%ID%", String.valueOf(port));
+                string = string.replace("%WORKER-ID%", workerId);
+                string = string.replace("%TOPOLOGY-ID%", stormId);
+                string = string.replace("%WORKER-PORT%", String.valueOf(port));
+                string = string.replace("%HEAP-MEM%", String.valueOf(memOnheap));
+                String[] strings = string.split("\\s+");
+                rets.addAll(Arrays.asList(strings));
+            }
+
+        } else if (value instanceof List) {
+            List<Object> objects = (List<Object>) value;
+            for (Object object : objects) {
+                String str = (String) object;
+                if (StringUtils.isNotBlank(str)){
+                    str = str.replace("%ID%", String.valueOf(port));
+                    str = str.replace("%WORKER-ID%", workerId);
+                    str = str.replace("%TOPOLOGY-ID%", stormId);
+                    str = str.replace("%WORKER-PORT%", String.valueOf(port));
+                    str = str.replace("%HEAP-MEM%", String.valueOf(memOnheap));
+                    rets.add(str);
+                }
+            }
+        }
+        return rets;
+    }
```

Comment:
It might be worth it to refactor this into two functions

```
private static String substituteChildOptsInternal(String string,  String workerId, String stormId, Long port, int memOnheap) {
  if (StringUtils.isNotBlank(string)){
    string = string.replace("%ID%", String.valueOf(port));
    string = string.replace("%WORKER-ID%", workerId);
    string = string.replace("%TOPOLOGY-ID%", stormId);
    string = string.replace("%WORKER-PORT%", String.valueOf(port));
    string = string.replace("%HEAP-MEM%", String.valueOf(memOnheap));
  }
  return string;
}

public List<String> substituteChildopts(Object value, String workerId, String stormId, Long port, int memOnheap) {
        List<String> rets = new ArrayList<>();
        if (value instanceof String) {
              String string = substituteChildOptsInternal((String) value,  workerId, stormId, port, memOnheap);
              if (StringUtils.isNotBlank(string)){
                String[] strings = string.split("\\s+");
                rets.addAll(Arrays.asList(strings));
              }
       } else if (value instanceof List) {
           List<Object> objects = (List<Object>) value;
           for (Object object : objects) {
               String str = substituteChildOptsInternal((String) object,  workerId, stormId, port, memOnheap);
               if (StringUtils.isNotBlank(str)){
                    rets.add(str);
               }
           }
       }
       return rets;
   }
```